lib/Mongo/MongoCommandCursor.php

```diff
@@ -112,4 +124,37 @@ public function __sleep()
     {
         return ['command'] + parent::__sleep();
     }
+
+    /**
+     * @see https://github.com/mongodb/mongo-php-driver-legacy/blob/1.6.14/db.c#L51
+     * @return bool
+     */
+    private function supportsReadPreference()
+    {
+        if ($this->command === []) {
+            return false;
+        }
+
+        $firstKey = array_keys($this->command)[0];
+        switch ($firstKey) {
+            case 'count':
+            case 'group':
+            case 'dbStats':
+            case 'geoNear':
+            case 'geoWalk':
+            case 'distinct':
+            case 'aggregate':
+            case 'collStats':
+            case 'geoSearch':
+            case 'parallelCollectionScan':
+                return true;
+
+            case 'mapreduce':
+            case 'mapReduce':
+                return !isset($this->command['out']) || isset($this->command['out']['inline']);
```

Comment:
If we're mimicking `ext-mongo` exactly ([here](https://github.com/mongodb/mongo-php-driver-legacy/blob/1.6.14/db.c#L96)), this should only return true when `out` is a PHP array and `inline` exists as an array key.

For what it's worth, both of `ext-mongo`'s checks are flawed. Ideally, `out` could be either an array or object and `inline` should probably have a true value (in the BSON sense, not PHP -- although that really depends on the server's mapReduce handling).

Reply:
I noticed the original check was flawed. Isn't omitting `out` the same as providing `out` with `inline` set? In that case, the absence of `out` would indicate the command supports readPreference since it's not required to run on a primary server to write results to a new collection.

That's why the check was written this way - when `out` isn't set or when `inline` is set (and ideally set to `true`, I agree on that) the command supports a readPreference. In all other cases we need to enforce a primary readPreference on the command.

Reply:
> Isn't omitting out the same as providing out with inline set?

Omitting `out` results in the following server error on 3.4: "out has to be a string or an object".

Moreover, the server doesn't seem to care about the value of `inline`, so even `{ inline: false }` is enough to produce an inline result.

So, if you'd like to mimic `ext-mongo` exactly, this can be changed to:

```php
return ! (isset($this->command['out']) &&
          is_array($this->command['out']) &&
          array_key_exists('inline', $this->command['out']));
```

And if you'd like a robust approach:

```
$out = isset($this->command['out']) ? $this->command['out'] : null;

return (is_array($out) || is_object($out)) && array_key_exists('inline', (array) $out);
```

Reply:
I've updated the code to mimic `ext-mongo`, even if it may be flawed.